### PR TITLE
Linregcleanup

### PR DESCRIPTION
**Pull request by cseed**
Builds on https://github.com/hail-is/hail/pull/2039

### CODE REVIEW

File: src/main/scala/is/hail/HailContext.scala [hunk 1]
```diff
@@ -449,12 +439,14 @@ class HailContext private(val sc: SparkContext,
     checkDatasetSchemasCompatible(vsms, inputs)
 
     // I can't figure out how to write this with existentials -cs
-    if (vsms(0).isGenericGenotype) {
-      val gdses = vsms.asInstanceOf[Array[GenericDataset]]
+    if (vsms(0).genotypeSignature == TGenotype) {
```
Comment (jbloom22) [inline src/main/scala/is/hail/HailContext.scala]: Should this be `!=`?

File: src/main/scala/is/hail/check/Gen.scala [hunk 2]
```diff
@@ -404,8 +406,12 @@ class Gen[+T](val gen: (Parameters) => T) extends AnyVal {
   // FIXME should be non-strict
   def withFilter(f: (T) => Boolean): Gen[T] = Gen { (p: Parameters) =>
     var x = apply(p)
-    while (!f(x))
+    var i = 0
+    while (!f(x)) {
+      assert(i < 100)
```
Comment (jbloom22) [inline src/main/scala/is/hail/check/Gen.scala]: It looks like you're doing rejection sampling that dies if unsuccessful after 100 attempts. Is the idea that our tests are seeded / deterministic so we need not worry about random failures?

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 3]
```diff
@@ -83,6 +83,7 @@ object RegressionUtils {
       }
       assert(completeSampleIndex(j) == i)
 
+
```
Comment (jbloom22) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: why extra line?

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 4]
```diff
@@ -281,6 +282,7 @@ object RegressionUtils {
       (yIS, covIS, vds.sampleIds)
         .zipped
         .filter((y, c, s) => y.forall(_.isDefined) && c.forall(_.isDefined))
+    println(yIS.map(_.toSeq), covIS.map(_.toSeq), vds.sampleIds)
```
Comment (jbloom22) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: println

File: src/test/scala/is/hail/methods/LinearRegressionSuite.scala [hunk 5]
```diff
@@ -10,16 +10,13 @@ import is.hail.variant.Variant
...
-  }
 
   def assertDouble(a: Annotation, value: Double, tol: Double = 1e-6) {
-    assert(D_==(a.asInstanceOf[Double], value, tol))
+    assert(D_==(a.asInstanceOf[IndexedSeq[Double]].apply(0), value, tol))
   }
 
   def assertEmpty(a: Annotation) {
-    assert(a == null)
+    // assert(a == null)
```
Comment (jbloom22) [inline src/test/scala/is/hail/methods/LinearRegressionSuite.scala]: You've left this for me to fix in my PR, right?

**Comment by jbloom22** [comment]
Looks like test failure is due to needing to wrap in covariates in array in this line
`top_5_pvals = (vds.linreg('sa.metadata.CaffeineConsumption')`
of
https://github.com/hail-is/hail/blob/master/python/hail/docs/tutorials/expression-language-part-2.ipynb

File: src/main/scala/is/hail/check/Gen.scala (hunk 2 above)
Comment (cseed) [inline src/main/scala/is/hail/check/Gen.scala]: If you filter too heavily, the tests will be slow (or, in the limit, go into an infinite loop).

File: src/main/scala/is/hail/HailContext.scala (hunk 1 above)
Comment (cseed) [inline src/main/scala/is/hail/HailContext.scala]: Yes, good catch.  This builds on https://github.com/hail-is/hail/pull/2039, I will fix it there.

File: src/main/scala/is/hail/check/Gen.scala (hunk 2 above)
Comment (jbloom22) [inline src/main/scala/is/hail/check/Gen.scala]: Cool, so you've added this to enforce efficient testing.

**Pull request merged**